RFC: Support arbitrary tagging and selection of testcases.

cdentyou opened the pull request:
Please consider these diffs for adding the ability selectively run tests based on arbitrary sets of tags.

The requirement is to provide a mechanism that will allow us to exclude test cases that need a long time to run from our normal unit test runs which we want to keep fast. We could of course use the existing mechanism for selecting by suite but currently we use suites to subdivide tests based on the functional area being tested so that a user can choose to just run the suite for one area.

In effect we want two orthogonal ways of selecting tests - by speed and by functional area.

We could subdivide the suites into long-running and slow running suites for each functional area but that would force the user (or some external test definition) to list all the suites according to the speed or the functional area that they want to test. Its also possible to envisage further orthogonal criteria that we might want in the future which will mean further lists will have to be maintained.

The alternative that we propose here is to allow an optional list of arbitrary tags (strings) to be provide when a testcase is being added to a suite and then to allow the srunner to take an include and exclude list of tags to allow it to filter testcases. (We would then register come testcases with a "SLOW_RUNNING" tag and run all tests excluding this tag in our normal UT).

brarcher commented:
To allow our other automated tests to run, I need to give our Jenkins instance the following command:

Jenkins: ok to test

You can disregard this message. I'm looking over the changes and will have a comment shortly.

brarcher commented on src/check.c:
```diff
@@ -166,7 +179,10 @@ static void tcase_free(TCase * tc)
     check_list_free(tc->unch_tflst);
     check_list_free(tc->ch_tflst);
-
+    if (tc->tags) {
```
nit pick, but please use the same style which is used elsewhere, namely place the { on the next line, and always use {} even if there is a one line body, for example:
```
if(condition)
{
   do_something();
}
```

brarcher commented on src/check.h.in:
```diff
@@ -170,6 +170,23 @@ CK_DLL_EXP void CK_EXPORT suite_add_tcase(Suite * s, TCase * tc);
+ *
+ * @return test case containing no tests
+ *
+ * @since 0.10.0
```
I'll be making the next release 0.11.0, so you can update this to say
```
@since 0.11.0
```

brarcher commented on src/check.h.in:
```diff
@@ -170,6 +170,23 @@ CK_DLL_EXP void CK_EXPORT suite_add_tcase(Suite * s, TCase * tc);
+ *
+ * @since 0.10.0
+ * */
+CK_DLL_EXP TCase *CK_EXPORT tcase_create_tagged(const char *name,
```
All test cases can have tags, it is just that the default state is to have a NULL list of tags, right? Why not add a method to add a tag to a test case?

As an example, every test case can have a timeout. This is applied using
```
tcase_set_timeout(TCase * tc, double timeout);
```
Perhaps it makes more sense to have a tagging function:
```
tcase_add_tag(TCase * tc, char* tag);
```

Is the use case such that adding multiple tags as once is desired? Should this really be:
```
tcase_add_tags(TCase * tc, char* tags);
```
or
```
tcase_add_tags(TCase * tc, char** tags);
```
? You are assuming that tags are space delimited. By convention this is true everywhere I have seen, but I attempted to find a "formal" definition for this and failed. Maybe it makes sense to have tags space delimited. I cannot think of a case where it does not make sense.

brarcher replied:
OK, I think I am fine with passing a string that contains multiple tags which are space delimited. 
What would be expected if a user invoked the call multiple times? Namely
```
tcase_add_tags(tc, "A B C");
tcase_add_tags(tc, "D");
```

Should the listed tags be: [A B C D]?

brarcher commented on src/check.h.in:
```diff
@@ -980,6 +997,7 @@ CK_DLL_EXP void CK_EXPORT srunner_run_all(SRunner * sr,
  */
 CK_DLL_EXP void CK_EXPORT srunner_run(SRunner * sr, const char *sname,
```
Making a change to this call will break the API for any user using this call. This call was [introduced in 2010](https://github.com/libcheck/check/commit/f84590bb43dfdf4364765c9c129f12b550440947).

The previous call will run a specific suite or test case. It may make more sense to add a separate call for running all tests for a given tag. Under the covers the new call and srunner_run() may call some other method which takes all the arguments for test case names, tags, etc.

cdentyou replied:
Ah yes - my bad - you pointed that out before but I lost it - will do

brarcher commented:
I'm leaving comments in the commit itself. As for the appveyor failure, appveyor is used to run builds and tests against Windows compilers. It seems that the check_check unit test program is dying due to a SEGFAULT:

https://ci.appveyor.com/project/brarcher46583/check/build/1.0.81/job/c1uktktsivnwjosy

> Test project C:/projects/check
>     Start 1: check_check
> 1/2 Test #1: check_check ......................***Exception: SegFault  0.91 sec

Could not be sure why without debugging it in Windows. I would suspect that if Windows is having an issue then maybe Valgrind may find something interesting in GNU/Linux or OSX.

brarcher commented:
Ah, the GNU/Linux tests which are run on the Jenkins instance are hitting a double free or corruption issue in glibc. See [here](https://check.ci.cloudbees.com/job/check-linux-github/50/COMPILER=gcc,FORK=no-fork,HOST=x86_64-redhat-linux,SNPRINTF=system-snprintf,SUBUNIT=no-subunit,TIMER=system-timer/console) for an example. 

> 09:29:23 check_check_tags.c:60:F:Black:black_test1:0: Black fail
> 09:29:23 *** glibc detected *** ./check_check: double free or corruption (out): 0x00000000019c7e20 ***
> 09:29:23 ======= Backtrace: =========
> 09:29:23 /lib64/libc.so.6(+0x7c00e)[0x7f0ae402a00e]
> 09:29:23 ./check_check[0x40cc71]
> 09:29:23 ./check_check[0x40b981]
> 09:29:23 ./check_check[0x41032c]
> 09:29:23 ./check_check[0x410389]
> 09:29:23 ./check_check[0x4103fb]
> 09:29:23 ./check_check[0x40fecb]
> 09:29:23 ./check_check[0x4107b3]
> 09:29:23 ./check_check[0x4106cf]
> 09:29:23 ./check_check[0x40c282]
> 09:29:23 /lib64/libc.so.6(__libc_start_main+0xf5)[0x7f0ae3fcf735]
> 09:29:23 ./check_check[0x401779]
> 09:29:23 ======= Memory map: ========
> 09:29:23 00400000-00419000 r-xp 00000000 00:31 1701363                            /scratch/jenkins/workspace/check-linux-github/COMPILER/gcc/FORK/no-fork/HOST/x86_64-redhat-linux/SNPRINTF/system-snprintf/SUBUNIT/no-subunit/TIMER/system-timer/tests/check_check
> 09:29:23 00618000-0061a000 rw-p 00018000 00:31 1701363                            /scratch/jenkins/workspace/check-linux-github/COMPILER/gcc/FORK/no-fork/HOST/x86_64-redhat-linux/SNPRINTF/system-snprintf/SUBUNIT/no-subunit/TIMER/system-timer/tests/check_check
> 09:29:23 019b7000-019d8000 rw-p 00000000 00:00 0                                  [heap]
> 09:29:23 7f0ae3d99000-7f0ae3dae000 r-xp 00000000 00:31 329091                     /usr/lib64/libgcc_s-4.7.2-20120921.so.1
> 09:29:23 7f0ae3dae000-7f0ae3fad000 ---p 00015000 00:31 329091                     /usr/lib64/libgcc_s-4.7.2-20120921.so.1
> 09:29:23 7f0ae3fad000-7f0ae3fae000 rw-p 00014000 00:31 329091                     /usr/lib64/libgcc_s-4.7.2-20120921.so.1
> 09:29:23 7f0ae3fae000-7f0ae415a000 r-xp 00000000 00:31 1597592                    /usr/lib64/libc-2.15.so
> 09:29:23 7f0ae415a000-7f0ae435a000 ---p 001ac000 00:31 1597592                    /usr/lib64/libc-2.15.so
> 09:29:23 7f0ae435a000-7f0ae435e000 r--p 001ac000 00:31 1597592                    /usr/lib64/libc-2.15.so
> 09:29:23 7f0ae435e000-7f0ae4360000 rw-p 001b0000 00:31 1597592                    /usr/lib64/libc-2.15.so
> 09:29:23 7f0ae4360000-7f0ae4365000 rw-p 00000000 00:00 0 
> 09:29:23 7f0ae4365000-7f0ae437b000 r-xp 00000000 00:31 1597618                    /usr/lib64/libpthread-2.15.so
> 09:29:23 7f0ae437b000-7f0ae457b000 ---p 00016000 00:31 1597618                    /usr/lib64/libpthread-2.15.so
> 09:29:23 7f0ae457b000-7f0ae457c000 r--p 00016000 00:31 1597618                    /usr/lib64/libpthread-2.15.so
> 09:29:23 7f0ae457c000-7f0ae457d000 rw-p 00017000 00:31 1597618                    /usr/lib64/libpthread-2.15.so
> 09:29:23 7f0ae457d000-7f0ae4581000 rw-p 00000000 00:00 0 
> 09:29:23 7f0ae4581000-7f0ae467b000 r-xp 00000000 00:31 1597600                    /usr/lib64/libm-2.15.so
> 09:29:23 7f0ae467b000-7f0ae487a000 ---p 000fa000 00:31 1597600                    /usr/lib64/libm-2.15.so
> 09:29:23 7f0ae487a000-7f0ae487b000 r--p 000f9000 00:31 1597600                    /usr/lib64/libm-2.15.so
> 09:29:23 7f0ae487b000-7f0ae487c000 rw-p 000fa000 00:31 1597600                    /usr/lib64/libm-2.15.so
> 09:29:23 7f0ae487c000-7f0ae4883000 r-xp 00000000 00:31 1597622                    /usr/lib64/librt-2.15.so
> 09:29:23 7f0ae4883000-7f0ae4a82000 ---p 00007000 00:31 1597622                    /usr/lib64/librt-2.15.so
> 09:29:23 7f0ae4a82000-7f0ae4a83000 r--p 00006000 00:31 1597622                    /usr/lib64/librt-2.15.so
> 09:29:23 7f0ae4a83000-7f0ae4a84000 rw-p 00007000 00:31 1597622                    /usr/lib64/librt-2.15.so
> 09:29:23 7f0ae4a84000-7f0ae4aa4000 r-xp 00000000 00:31 1597585                    /usr/lib64/ld-2.15.so
> 09:29:23 7f0ae4c43000-7f0ae4c87000 rw-p 00000000 00:00 0 
> 09:29:23 7f0ae4ca0000-7f0ae4ca3000 rw-p 00000000 00:00 0 
> 09:29:23 7f0ae4ca3000-7f0ae4ca4000 r--p 0001f000 00:31 1597585                    /usr/lib64/ld-2.15.so
> 09:29:23 7f0ae4ca4000-7f0ae4ca5000 rw-p 00020000 00:31 1597585                    /usr/lib64/ld-2.15.so
> 09:29:23 7f0ae4ca5000-7f0ae4ca6000 rw-p 00000000 00:00 0 
> 09:29:23 7ffd235e6000-7ffd23607000 rw-p 00000000 00:00 0                          [stack]
> 09:29:23 7ffd237f2000-7ffd237f4000 r-xp 00000000 00:00 0                          [vdso]
> 09:29:23 ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
> 09:29:23 /bin/sh: line 5: 20096 Aborted                 (core dumped) ${dir}$tst
> 09:29:23 FAIL: check_check

brarcher commented:
You may be able to reproduce the issue locally if you disable using fork(). It appears that the issue manifests when not using fork(), which is the case for Windows and the failed tests on GNU/Linux.

Try building with --no-fork, or running a simplified unit test example program with CK_FORK=no in the environment.

brarcher commented on tests/check_check_tags.c:
```diff
@@ -0,0 +1,315 @@
+#include <assert.h>
+#include "check_check.h"
+
+/* from check_check_master.c */
```
It may be better to add these tests to the same places where the other tests are located. Try in check_check_selective.c. Check allows a unit test to create suites and run unit tests. This is how Check tests itself. You may need to do this in check_check_selective.c.

What may get you is that all the suites which are added in check_check_main.c have an entry in check_check_master.c in the following table:
```
static master_test_t master_tests[]
```
where the entry specifies the suite name, the expected result, and the expected output. For example:
```
{ "Simple Tests", CK_FAILURE, "This test should fail" },
```
Each suite is listed in the order they are added, then for each suite the tests are in the order they were added.

Hopefully by adding the tests to the master list some of the duplicated code from check_check_master.c can be avoided.

brarcher commented on src/check_run.c:
```diff
@@ -158,8 +159,47 @@ static void srunner_run_end(SRunner * sr,
+ * Helper func to compare two lists of tags and return true if there is a
+ * common tag.
+ */
+static unsigned int matching_tag(const char *tags1_orig, const char *tags2_orig)
```
Why not handle the splitting of tags from a string into a list in tcase_add_tags()? That may make this easier.

brarcher commented on src/check_run.c:
```diff
@@ -158,8 +159,47 @@ static void srunner_run_end(SRunner * sr,
+	return 0;
+
+    tags1 = strdup(tags1_orig);
+    tag1 = strtok(tags1, " ");
```
Can strtok() be used in the way it is being used here? Namely, strtok() maintains state on its original string. However, strtok is being used here to maintain state on two different strings, namely tags1_orig and tags2_orig. Should strtok_r be used instead to avoid issues?

cdentyou replied:
Yes - strtok_r with different placeholder pointers is what's needed here.

brarcher commented on tests/check_check_tags.c:
```diff
@@ -0,0 +1,315 @@
+
+  n_expected_msgs = sizeof(expected_msgs)/sizeof(expected_msgs[0]);
+
+  srunner_run(sr, NULL, NULL, NULL, NULL, CK_VERBOSE);
```
I would like to see a test where an empty include tag is passed, or tags which do not exist are passed, and thus no tests are run. There should also be tests for the exclusion tags as well. 

Also, what happens if one separates tags with multiple spaces? Can there be a test for this as well?

Also, what if someone passes "Yellow" as both an included and excluded test? What should happen? Is this defined? Can there be a test for this as well?

brarcher commented:
I think that the idea of tagging tests and running tests based on tags has merit. Lets flush this out a bit more and see where it goes.

cdentyou commented:
Interim push to get automated test results - fixed double free crash, wrong use of strtok rather than strtok_r, version number in func comment.
Code format, additional UTs and API changes still to come ..

brarcher commented:
It now appears that when running without fork() that the following test is failing:
```
05:21:04 Running suite(s): Check Tag Filtering
05:21:04 0%: Checks: 3, Failures: 3, Errors: 0
05:21:04 check_check_tags.c:54:F:Yellow:yellow_test1:0: Yellow fail
05:21:04 check_check_tags.c:36:F:Red:red_test1:0: Red fail
05:21:04 check_check_tags.c:48:F:Purple:purple_test1:0: Purple fail
...
05:21:04 check_check_tags.c:251:F:include tags:inc_red:0: Expected to run 2 tests but actually ran 3
```
However when fork() is used it is running the expected number of tests:
```
05:57:45 Running suite(s): Check Tag Filtering
05:57:45 0%: Checks: 2, Failures: 2, Errors: 0
05:57:45 check_check_tags.c:36:F:Red:red_test1:0: Red fail
05:57:45 check_check_tags.c:48:F:Purple:purple_test1:0: Purple fail
...
05:57:46 check_check_tags.c:257:P:include tags:inc_red:0: Passed
```
The issue is probably something simple, but I cannot see it at the moment. As the issue is occurs without using fork(), it is likely something about a data structure being used during a test that is not being reset as it should, but when the test runs with fork() the structure modifications are being discarded when the separate test memory space is discarded.

cdentyou commented:
Hopefully getting closer. I think the main thing not done is undoing the API change to srunner_run (see question on that). 

Fixed sigsegv.

Fixed all the layout convention violations that I can see (is there a check patch like tool I can use ?)

Changed layout of tags inside test case to be an array of strings - mostly to avoid the nested use of strtok (since windows doesn't seem to have strtok_r).

Provided set/get tags on TCase and got rid of special create function.
 
Added tests to verify set/get.
 
I have also reworked the tag tests in line with what check_check_selective does which has made them much simpler (thanks !). Have not combined into that file as it might make that file rather long - but don't mind.

Extended tests to combine include and exclude and to do various 'silly' combinations.

brarcher commented on src/check_impl.h:
```diff
@@ -65,6 +65,8 @@ struct TCase
     List *ch_sflst;
     List *ch_tflst;
+    unsigned int num_tags;
+    char **tags;
```
Check has a list structure available in [check_list.c](https://github.com/libcheck/check/blob/master/src/check_list.c). Consider using that instead of an char**.

cdentyou replied:
OK - will take a look at the list stuff - wanted to get the self tests all passing reliably first - but touch wood that seems to have happened.

brarcher commented on src/check.h.in:
```diff
@@ -170,6 +170,41 @@ CK_DLL_EXP void CK_EXPORT suite_add_tcase(Suite * s, TCase * tc);
+ *
+ * @since 0.11.0
+ * */
+CK_DLL_EXP int CK_EXPORT tcase_get_tags(TCase * tc,
```
I'm not sure that a getter for the tags is necessary. The other calls in this API which setup unit tests are not symmetrical, and expect one to set but not inspect unit test structures.

cdentyou replied:
Its certainly not absolutely necessary.

It could be useful if anybody has a use case for inserting extra tags (or removing existing tags) against tests that had already been tagged.  Its probably not that likely but there's also a slight benefit when it comes to the unit tests as I can isolate testing of set/get from the test runner functionality.

If you would prefer it removed I'm fine with that but let me know.

brarcher replied:
Unless you plan on using this call yourself, I would prefer it removed. I think it adds a bit of extra complexity. That, and at least my typical usecase is to setup tests then run them as opposed to modifying them at runtime.

brarcher commented on src/check.h.in:
```diff
@@ -170,6 +170,41 @@ CK_DLL_EXP void CK_EXPORT suite_add_tcase(Suite * s, TCase * tc);
+ *
+ * @param tc the test case
+ *
+ * @param tags string containing arbitrary tags separated by spaces
```
... separated by spaces**.**

Missing the '.' at the end

cdentyou replied:
done.

brarcher commented on src/check.c:
```diff
@@ -149,10 +150,113 @@ TCase *tcase_create(const char *name)
+    tag = strtok(tags, " ");
+    while (tag)
+    {
+	if (strlen(tag))
```
Would is ever be true that the length of a tag is zero? Would not strtok return NULL instead?

cdentyou replied:
Yes I think you are right - pure paranoia on my part. Since it clutters the code I will remove the redundant length check.

brarcher commented on src/check.c:
```diff
@@ -166,10 +270,40 @@ static void tcase_free(TCase * tc)
+    
+    tags = strdup(check_for);
+
+    tag = strtok(tags, " ");
```
Is there was a helper method which converted a string into a list of tokens then this parsing would only need to be done once, instead of here and in tcase_set_tags()

cdentyou replied:
Agreed. Will take a look at this alongside changing to use the list.

brarcher commented on src/check.c:
```diff
@@ -166,10 +270,40 @@ static void tcase_free(TCase * tc)
+	{
+	    if (0 == strcmp(tc->tags[i], tag))
+	    {
+		return 1;
```
does returning here leak tags, which contains strdup'ed data?

cdentyou replied:
Good catch ! fixed.

brarcher commented on tests/check_check_tags.c:
```diff
@@ -0,0 +1,621 @@
+
+START_TEST(red_test1)
+{
+    test_executed[RED] = 1;
```
As each of the tests may be run in its own memory space due to using fork(), I'm not sure that this will work as expected. Using the asserts was probably a better way of checking if a specific test was run or not.

cdentyou replied:
I think what I'm doing here is the same as is used by check_check_selective - I agree it depends on forking being disabled but in both selective_setup() and tag_test_setup() the fork_status is explicitly set to CK_NOFORK.

My question is whether limiting the testing to NOFORK mode matters ? I feel I have enough of an overview to judge that so I'll be guided by you. 

In terms of the code it would be easy enough for me to go back to the assert driven scheme but in my view there is a slight readability and code size cost to that approach - and I should probably do some tweaks to check_check_main/master to properly share the pieces of code I need to help handle the logging.

brarcher replied:
After reading through the tests, I think they are fine. I did not realize that you were setting NOFORK for them.

I think that the main/master stuff is useful if tests are expected to fail, and the failure message themselves need to be checked. If your tests are all expected to pass, and do different inspections to determine that the expected tests were run, I think that is fine.

brarcher commented:
Things are coming along, keep it up! I have not gone through the unit tests you proposed yet, but will do that shortly.

In addition to the tests you have here, there are a few other types of tests which should be added to ensure proper coverage.

- [check_mem_leaks.c](https://github.com/libcheck/check/blob/master/tests/check_mem_leaks.c) is a separate program which exercises each public API at least once, and is run under valigrind to ensure that there are no memory leaks. You may be able to add your suite to this test and have it run. The script which run the test under valigrind is [test_mem_leaks.sh](https://github.com/libcheck/check/blob/master/tests/test_mem_leaks.sh).
- The script [test_output.sh](https://github.com/libcheck/check/blob/master/tests/test_output.sh) runs an example program [ex_output.c](https://github.com/libcheck/check/blob/master/tests/ex_output.c) with multiple different configurations. It accepts some command line arguments which change what parameters are set for the unit tests. In addition, it is used to check that environment variables are honored. There should be a sequence of tests in this file checking that passing different tags to suites which may or may not have tags works as expected, as well as setting the proposed environment variables CK_INC_TAGS and CK_EXC_TAGS. It will also need to test that the environment variables override whatever was explicitly set in the code.

The environment variables maybe should be spelled out instead of abbreviated. The other environment variables summarized [here](https://libcheck.github.io/check/doc/check_html/check_7.html#Environment-Variable-Reference) are not abbreviated, except for the CK_ part.

Finally, some documentation will need to be added once all it said and done. The main documentation file is located [here](https://github.com/libcheck/check/blob/master/doc/check.texi). Here is the section of the generated documentation which may need to mention the [new API calls](https://libcheck.github.io/check/doc/check_html/check_4.html#Selective-Running-of-Tests) and environment variables, and [here](https://libcheck.github.io/check/doc/check_html/check_7.html#Environment-Variable-Reference) is a summary of the environment variables.

brarcher commented on tests/check_check_tags.c:
```diff
@@ -0,0 +1,621 @@
+
+    for (i = 0; i < MAX_TESTS; i++)
+    {
+	test_executed[i] = 0;
```
Does this reset all the tests to RED? To let the default value be not an option, should the enum instead be initialized as:

```
enum
{
   RED = 1,
...
```

Also, why not just do a memset() instead?

cdentyou replied:
test_executed[test_id] is really a boolean - 1 for if the test <test_id> has run and 0 if it hasn't. This function sets the flags for all tests to false.

Agree that using memset is more succinct so changing to that.

BTW am I right that you don't have a boolean type defined anywhere in the check code ?

brarcher commented on tests/check_check_tags.c:
```diff
@@ -0,0 +1,621 @@
+
+    s = suite_create("Check Tag Filtering");
+
+    red = tcase_create("Red");
```
nitpick, but why not initialize the Red tests in the same was as the other tests? Namely, 4 lines together without blank lines?

cdentyou replied:
Good catch - that was really annoying. Fat fingered moment I guess. Now fixed.

brarcher commented on tests/check_check_tags.c:
```diff
@@ -0,0 +1,621 @@
+
+    tc = tcase_create("TC");
+    tcase_set_tags(tc, expected_str);
+    ret_val = tcase_get_tags(tc, ret_str, sizeof(ret_str));
```
Not sure that a tcase_get_tags() call is necessary. Should this be removed from the API? Do you foresee a compelling usecase for it?

cdentyou replied:
As I said in comment next to the function. Use case is if anybody ends up with a test architecture where tagging is incrementally applied/removed on tests that have already been declared. I don't think I have that case today but it seems sort of logical to provide the functionality given that we have separate the 'set_tags' operation from the create operation.

Also failures in tests like set_get_empty/set_get_null are going to be easier to diagnose this way than if they have to see what tests are actually run. 

But these are pretty minor benefits ...

brarcher commented on tests/check_check_tags.c:
```diff
@@ -0,0 +1,621 @@
+    ck_assert(test_executed[YELLOW]);
+    ck_assert(test_executed[BLACK]);
+
+    reset_executed();
```
If the state is asserted at the beginning of the test, why bother doing so at the end also?

cdentyou replied:
paranoia - removed.

brarcher commented on tests/check_check_tags.c:
```diff
@@ -0,0 +1,621 @@
+
+} END_TEST
+
+START_TEST(exc_nothing)
```
can you spell out the abbreviations for inc and exc? Initially when I looked at "exc" I misread it as "exe" and thought "the test executes nothing" which was incorrect.

cdentyou replied:
just caught me out too - done.

brarcher commented on tests/check_check_tags.c:
```diff
@@ -0,0 +1,621 @@
+} END_TEST
+
+
+START_TEST(inc_red_exc_red)
```
So, the understanding of the include and exclude parameters is that the exclude will always prevail? Just make this clear in the documentation.

cdentyou replied:
Done. I'm not 100% sure which way around is better but this seems fairly logical.

brarcher commented on src/check.c:
```diff
@@ -149,10 +150,113 @@ TCase *tcase_create(const char *name)
+	return;
+    }
+    
+    tc->tags = calloc(count, sizeof(const char *));
```
You will need to do a cast here to avoid an error from clang:
```
 check.c:202:14: fatal error: assigning to 'char **' from incompatible type 'void *'; 
    tc->tags = calloc(count, sizeof(const char *));
```

cdentyou replied:
done

cdentyou commented:
Builds and Tests seem to be all passing - only thing left is looking at using check_lists instead of char ** for storing tags against test case.

cdentyou commented:
OK - latest push addresses all comments (I think) apart from the removal of tcase_get_tags() which can be done if still desired.
Note includes a commit to change one of the timeout tests because it was repeatedly failing for me - tweaking it so that there's an extra .1 second leeway seems to have made it consistently pass.

cdentyou commented:
So sadly looks like the self test failure has come back - I think everything stems from the failure of test 97 which I think boils down the test at line 648 passing instead of timing out. But with my change the timeout should be 1.5 seconds and the test sleeps for 2.  Can't see how I can have caused this but will keep hunting. I also need to fix the test runner API breakage.


Users/travis/build/libcheck/check/tests/check_check_master.c:271:F:Core Tests:test_check_nfailures:0: Unexpected number of failures received, 100, expected 101.
/Users/travis/build/libcheck/check/tests/check_check_master.c:278:P:Core Tests:test_check_ntests_run:0: Passed
/Users/travis/build/libcheck/check/tests/check_check_master.c:348:F:Core Tests:test_check_failure_msgs:0: For test 109: Expected Test timeout expired, got Early exit with return value 1
/Users/travis/build/libcheck/check/tests/check_check_master.c:414:F:Core Tests:test_check_failure_ftypes:0: Failure type wrong for test 110
/Users/travis/build/libcheck/check/tests/check_check_master.c:387:F:Core Tests:test_check_failure_lnos:0: For test 97 (failure 82): Expected lno 648, got 661 for suite Timeout Double Scaling Tests, msg Test timeout expired

brarcher commented on src/check.c:
```diff
@@ -166,10 +206,34 @@ static void tcase_free(TCase * tc)
+	for(check_list_front(tc->tags); !check_list_at_end(tc->tags);
+	    check_list_advance(tc->tags))
+	{
+	    if (0 == strcmp(check_list_val(tc->tags), check_list_val(check_for)))
```
Clang did not like this:
```
16:15:47 check.c:228:15: fatal error: no matching function for call to 'strcmp'
16:15:47             if (0 == strcmp(check_list_val(tc->tags), check_list_val(check_for)))
16:15:47                      ^~~~~~
16:15:47 /usr/include/string.h:143:12: note: candidate function not viable: cannot convert argument of incomplete type 'void *' to 'const char *'
16:15:47 extern int strcmp (__const char *__s1, __const char *__s2)
16:15:47            ^
```
Might need to cast the contents of the list item to char*

brarcher commented on src/check.c:
```diff
@@ -149,10 +150,49 @@ TCase *tcase_create(const char *name)
     tc->ch_sflst = check_list_create();
     tc->unch_tflst = check_list_create();
     tc->ch_tflst = check_list_create();
+    tc->tags = NULL;
```
The other lists are created and destroyed unconditionally, Why not follow this pattern? Or, is there an advantage to doing it this way?

cdentyou replied:
Hi Branden,

The only reason was that the tag list is always replaced rather than being
modified so the initial list would end up being freed .. but agree that
always having a list removes a special case .. will remove.

Thx

Cris


On Fri, Jun 24, 2016 at 10:45 PM, Branden Archer <notifications@github.com>
wrote:

> In src/check.c
> <https://github.com/libcheck/check/pull/44#discussion_r68467730>:
>
> > @@ -149,10 +150,49 @@ TCase *tcase_create(const char *name)
> >      tc->ch_sflst = check_list_create();
> >      tc->unch_tflst = check_list_create();
> >      tc->ch_tflst = check_list_create();
> > +    tc->tags = NULL;
>
> The other lists are created and destroyed unconditionally, Why not follow
> this pattern? Or, is there an advantage to doing it this way?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/libcheck/check/pull/44/files/a9cd88e4e4b0864e4c416c15c78e5e06c07da514..e3f6eb3ed7044f28d835e1c063aad5d448f6be3e#r68467730>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ALhg7a6X_7ySEfYiLetqQoUuo2S2TFypks5qPFAUgaJpZM4IuPMy>
> .
>

brarcher commented on src/check.h.in:
```diff
@@ -987,6 +1027,33 @@ CK_DLL_EXP void CK_EXPORT srunner_run(SRunner * sr, const char *sname,
+ * Overrides any include criteria.
+ * @param print_mode the verbosity in which to report results to stdout
+ *
+ * @since 0.9.9
```
```
@since 0.9.9
```
This may need to be changed to 0.11.0

brarcher commented on src/check.h.in:
```diff
@@ -977,6 +1012,11 @@ CK_DLL_EXP void CK_EXPORT srunner_run_all(SRunner * sr,
+ * that share one of these tags. A NULL means "any test case".
+ * @param exclude_tags space separate list of tags. Only run test cases
+ * that do not share one of these tags. A NULL means "any test case".
+ * Overrides any include criteria.
```
You do not need these comments anymore, right?

brarcher replied:
Oh, never mind. I thought the comments were over another function. These are fine.

brarcher commented on tests/check_check_tags.c:
```diff
@@ -0,0 +1,413 @@
+/*
```
I like the tests in this file. Very thorough.

brarcher commented:
> So sadly looks like the self test failure has come back - I think everything stems from the failure of test 97 which I think boils down the test at line 648 passing instead of timing out. But with my change the timeout should be 1.5 seconds and the test sleeps for 2

I think that the VMs running on Travis-CI are not always as responsive as one would hope. Some of the tests by their nature assume a certain amount of responsiveness to meet timing deadlines or detect failed deadlines. I think your change to the test timing is OK, as it does not break the spirit of the test and makes the window for detecting the timed out test case wider. I would not worry too much about that sporadic failure.

brarcher commented:
> OK - latest push addresses all comments (I think)

Kindly remember to take a look at the check_mem_leaks test and the test_output.sh test. Both are mentioned in a previous comment, look there for details.

>  apart from the removal of tcase_get_tags() which can be done if still desired.

If you have a need for this call in your personal use case of tags, then keep it. Otherwise, I think we should remove it, as it would be additional code to test and maintain going forward.

cdentyou commented:
Hi Branden

I will take a look at the check_mem_leaks test and make sure I've removed
the _get_ function tomorrow - many thanks for all your help so far - I
really appreciate it - I'm a bit slow because stuff keeps coming up in the
day job but I should have some time to concentrate on this now.

Thanks

Cris

On Fri, Jun 24, 2016 at 11:09 PM, Branden Archer <notifications@github.com>
wrote:

> OK - latest push addresses all comments (I think)
>
> Kindly remember to take a look at the check_mem_leaks test and the
> test_output.sh test. Both are mentioned in a previous comment, look there
> for details.
>
> apart from the removal of tcase_get_tags() which can be done if still
> desired.
>
> If you have a need for this call in your personal use case of tags, then
> keep it. Otherwise, I think we should remove it, as it would be additional
> code to test and maintain going forward.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/libcheck/check/pull/44#issuecomment-228474549>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ALhg7dTCSWu_8NHlInaMFyeTLa-DcBW3ks5qPFWxgaJpZM4IuPMy>
> .
>

brarcher commented:
> many thanks for all your help so far

No problem. (:

> stuff keeps coming up in the day job

The day job is more important, no worries. When you have time this feature can be wrapped up and submitted. There is no rush.

cdentyou commented:
Added tags tests into check_em_leaks - running check_mem_leaks.sh seems to result in a pass. 
Fixed comments in check.h, made tags list always exist ..
Will go over the comments one more time to see if there's anything else I need to do but right now I think that's it ...

brarcher commented on README:
```diff
@@ -0,0 +1,55 @@
+# About Check
```
Did not realize it until now, but your commit is adding a lot of files. Most of them are probably generated at build time and should not be commited.

cdentyou replied:
Oops - I see what you mean - I'll go and clean those out - not sure how that happened - wrong git add option I suspect.

brarcher commented:
valgrind is indicating that there are [some memory leaks](https://check.ci.cloudbees.com/job/check-linux-github/70/COMPILER=gcc,FORK=fork,HOST=x86_64-redhat-linux,SNPRINTF=system-snprintf,SUBUNIT=no-subunit,TIMER=system-timer/consoleFull). For example:
```
12:21:03 ==8506== 192 (144 direct, 48 indirect) bytes in 6 blocks are definitely lost in loss record 295 of 299
12:21:03 ==8506==    at 0x4C2981C: malloc (vg_replace_malloc.c:270)
12:21:03 ==8506==    by 0x4E3451E: emalloc (in /scratch/jenkins/workspace/check-linux-github/COMPILER/gcc/FORK/fork/HOST/x86_64-redhat-linux/SNPRINTF/system-snprintf/SUBUNIT/no-subunit/TIMER/system-timer/src/.libs/libcheck.so.0.0.0)
12:21:03 ==8506==    by 0x4E34619: check_list_create (in /scratch/jenkins/workspace/check-linux-github/COMPILER/gcc/FORK/fork/HOST/x86_64-redhat-linux/SNPRINTF/system-snprintf/SUBUNIT/no-subunit/TIMER/system-timer/src/.libs/libcheck.so.0.0.0)
12:21:03 ==8506==    by 0x4E3368F: tag_string_to_list (in /scratch/jenkins/workspace/check-linux-github/COMPILER/gcc/FORK/fork/HOST/x86_64-redhat-linux/SNPRINTF/system-snprintf/SUBUNIT/no-subunit/TIMER/system-timer/src/.libs/libcheck.so.0.0.0)
12:21:03 ==8506==    by 0x4E33768: tcase_set_tags (in /scratch/jenkins/workspace/check-linux-github/COMPILER/gcc/FORK/fork/HOST/x86_64-redhat-linux/SNPRINTF/system-snprintf/SUBUNIT/no-subunit/TIMER/system-timer/src/.libs/libcheck.so.0.0.0)
12:21:03 ==8506==    by 0x4E33861: tcase_free (in /scratch/jenkins/workspace/check-linux-github/COMPILER/gcc/FORK/fork/HOST/x86_64-redhat-linux/SNPRINTF/system-snprintf/SUBUNIT/no-subunit/TIMER/system-timer/src/.libs/libcheck.so.0.0.0)
12:21:03 ==8506==    by 0x4E3346B: suite_free (in /scratch/jenkins/workspace/check-linux-github/COMPILER/gcc/FORK/fork/HOST/x86_64-redhat-linux/SNPRINTF/system-snprintf/SUBUNIT/no-subunit/TIMER/system-timer/src/.libs/libcheck.so.0.0.0)
12:21:03 ==8506==    by 0x4E33F95: srunner_free (in /scratch/jenkins/workspace/check-linux-github/COMPILER/gcc/FORK/fork/HOST/x86_64-redhat-linux/SNPRINTF/system-snprintf/SUBUNIT/no-subunit/TIMER/system-timer/src/.libs/libcheck.so.0.0.0)
12:21:03 ==8506==    by 0x401B84: main (in /scratch/jenkins/workspace/check-linux-github/COMPILER/gcc/FORK/fork/HOST/x86_64-redhat-linux/SNPRINTF/system-snprintf/SUBUNIT/no-subunit/TIMER/system-timer/tests/.libs/lt-check_mem_leaks)
```

Namely, when a test case is freed a new list is created. This path only occurs when using fork(), so compiling check with --disable-fork [would not have found this](https://check.ci.cloudbees.com/job/check-linux-github/70/COMPILER=gcc,FORK=no-fork,HOST=x86_64-redhat-linux,SNPRINTF=system-snprintf,SUBUNIT=no-subunit,TIMER=system-timer/console).

Clang is finding the following it is [unhappy](https://check.ci.cloudbees.com/job/check-linux-github/70/COMPILER=clang++,FORK=no-fork,HOST=x86_64-redhat-linux,SNPRINTF=system-snprintf,SUBUNIT=no-subunit,TIMER=system-timer/console) about:
```
12:00:47 :228:15: fatal error: no matching function for call to 'strcmp'
12:00:47             if (0 == strcmp(check_list_val(tc->tags), check_list_val(check_for)))
12:00:47                      ^~~~~~
12:00:47 /usr/include/string.h:143:12: note: candidate function not viable: cannot convert argument of incomplete type 'void *' to 'const char *'
12:00:47 extern int strcmp (__const char *__s1, __const char *__s2)
```
This will need to be corrected, as it is resulting in a build failure.

cdentyou commented:
The memleak you point out makes sense to me - its actually fallout from the change I just made to always have a list - 'always' ended up including the test case free code.
Where did you find that dump of the leak - I can see that memchecks is failing in the test logs but not nice callstacks like you had ...

brarcher commented:
> Where did you find that dump of the leak - I can see that memchecks is failing in the test logs but not nice callstacks like you had

The link I mentioned in the previous comment points to the build output from one of the tests on the Jenkins instance. If you look for the string "are definitely lost" you will find valgrind report stack traces.

In addition, you may see lots of SEGV messages in the logs. Those will not affect the test much, but are annoying and are from a bug I introduced unknowingly a few years back. This is now fixed in master, if you want to rebase your changes.

cdentyou commented:
I'm not sure what my Travis failures mean - MacOS builds seem to be silently timing out. I notice that the last build of master seems to have similar problems with the MacOS builds so I'm wondering if this is something I've picked up from master ? I can't seem to find more information than this message - 

"No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated"

It looks like the build is being kicked off but does not produce any output ...

brarcher commented:
> I'm not sure what my Travis failures mean - MacOS builds seem to be silently timing out.

I'm not sure where to see this. Can you provide a link to an example?

> I notice that the last build of master seems to have similar problems with the MacOS builds so I'm wondering if this is something I've picked up from master ?

The last several tests runs against master were successful according to [this](https://travis-ci.org/libcheck/check/builds). However, sometimes there are sporadic timeouts on OSX, such as [this one](https://travis-ci.org/libcheck/check/jobs/135500062) 29 days ago.

> I can't seem to find more information than this message 

I've not seen that message, but I wonder if it is from Travis-CI itself, attempting to abort potentially hung jobs after a timeout.

The most recently change set did result in a failure on the Jenkins-based tests. I think maybe it expected something to time out that did not. Let me see if it happens again.

Jenkins: test this please

brarcher commented on src/check_impl.h:
```diff
@@ -65,6 +65,8 @@ struct TCase
     List *unch_tflst;
     List *ch_sflst;
     List *ch_tflst;
+    unsigned int num_tags;
```
I do not think that this is used anywhere. It is still needed?

brarcher commented:
I think that I like the changes. There was only one small comment I just posted about a variable not being needed anymore. Besides that, things are looking good.

As these changes add an environment variable, there is one more test which will need to be added which verifies the environment variable usage. This was mentioned earlier, but may have been lost in the discussion. I'll copy the mention below:

> The script [test_output.sh](https://github.com/libcheck/check/blob/master/tests/test_output.sh) runs an example program [ex_output.c](https://github.com/libcheck/check/blob/master/tests/ex_output.c) with multiple different configurations. It accepts some command line arguments which change what parameters are set for the unit tests. In addition, it is used to check that environment variables are honored. There should be a sequence of tests in this file checking that passing different tags to suites which may or may not have tags works as expected, as well as setting the proposed environment variables CK_INC_TAGS and CK_EXC_TAGS. It will also need to test that the environment variables override whatever was explicitly set in the code.

Finally, the environment variables will need to be documented. Here is a description from earlier in the discussion:

> Finally, some documentation will need to be added once all it said and done. The main documentation file is located [here](https://github.com/libcheck/check/blob/master/doc/check.texi). [Here](https://libcheck.github.io/check/doc/check_html/check_4.html#Selective-Running-of-Tests) is the section of the generated documentation which may need to mention the new API calls and environment variables, and [here](https://libcheck.github.io/check/doc/check_html/check_7.html#Environment-Variable-Reference) is a summary of the environment variables.

cdentyou commented:
"As these changes add an environment variable, there is one more test which will need to be added which verifies the environment variable usage. This was mentioned earlier, but may have been lost in the discussion. I'll copy the mention below: ..."

I added some tests to check_check_tags.c that are driven by the environment variables in the same way that check_check_selective.c does it for CK_RUN_SUITE/CASE.  I took at look at test_output.sh and ex_output.c but it wasn't obvious to me what I needed to do there - they don't seem to do anything related to CK_RUN_SUITE/CASE either. 

"Finally, some documentation will need to be added once all it said and done. The main documentation file is located here. Here is the section of the generated documentation which may need to mention the new API calls and environment variables, and here is a summary of the environment variables."

I have added a section into the documentation which make sense to me - but would appreciate feedback from fresh eyes. And I've added the env vars to the appendix and matched the indexing that I see for CK_RUN_SUITE ...

Let me know if there's anything else

brarcher commented:
> I added some tests to check_check_tags.c that are driven by the environment variables in the same way that check_check_selective.c does it for CK_RUN_SUITE/CASE.

I see. You are right, that should be sufficient. 

> I took at look at test_output.sh and ex_output.c but it wasn't obvious to me what I needed to do there - they don't seem to do anything related to CK_RUN_SUITE/CASE either.

Oh, I did not realize there were no tests for CK_RUN_* there. However, as the same coverage is provided elsewhere, no point in duplicating the effort.

> I have added a section into the documentation which make sense to me - but would appreciate feedback from fresh eyes.

I'll take a look

brarcher commented on doc/check.texi:
```diff
@@ -1314,8 +1316,70 @@ two environment variables available that offers this ability,
+
+It can also be useful to be able to dynamically select or exclude
+groups of tests to be run based on criteria other than the suite or
+testcase name. For example certain test cases can be tagged as
```
testcase => test case

cdentyou replied:
done

brarcher commented on doc/check.texi:
```diff
@@ -1301,11 +1302,12 @@ srunner_add_suite (sr, make_pack_suite ());
 @end example
 
-@node Selective Running of Tests, Testing Signal Handling and Exit Values, Multiple Suites in one SRunner, Advanced Features
+@node Selective Running of Tests, Selecting Tests Based on Arbitrary Tags, Multiple Suites in one SRunner, Advanced Features
```
This may be too picky, but I feel that the section covering selectively running tests should cover both the idea of a suite or test case as well as tagging. That is, the following would be the sections:
```
4.6 Selective Running of Tests
4.6.1 Selecting Tests by Suite or Test Case
4.6.2 Selecting Tests by Tags
```

Let me give a stab at how this may flow:

```
4.6 Selective Running of Tests
After adding a couple of suites and some test cases in each, it is sometimes practical to 
be able to run only one suite, or one specific test case, without recompiling the test code. 
Check provides two ways to accomplish this, either by specifying a suite or test case by name
or by assigning tags to test cases and specifying one or more tags to run.

4.6.1 Selecting Tests by Suite or Test Case
There are two environment variables available that can be used to run a specific suite or
test case by name: CK_RUN_SUITE and CK_RUN_CASE. Just set the value to the name
of the suite and/or test case you want to run. These environment variables can also be a 
good integration tool for running specific tests from within another tool, e.g. an IDE.

4.6.2 Selecting Tests by Tags
(the content you wrote).
```

cdentyou replied:
Makes sense. Slight delay while I figure out how to do subsections .. but I'll do that.

brarcher commented on doc/check.texi:
```diff
@@ -1314,8 +1316,70 @@ two environment variables available that offers this ability,
+not be run. In cases where both @code{CK_INCLUDE_TAGS} and
+@code{CK_EXCLUDE_TAGS} match then the test will be excluded.
+
+Both @code{CK_INCLUDE_TAGS} and @code{CK_EXCLUDE_TAGS} can be
```
I was wondering about this. Could you add a test which verifies this behavior?

cdentyou replied:
You mean one that explictly uses the env variables ? There is already an include_exclude suite that tests the correct filtering - its just that it does not use the env vars - it would certainly be easy enough to clone it to produce an env var version as well if you think that's useful. Probably should just for completeness.

brarcher replied:
Some of the tests in check_check_tags.c do verify that CK_INCLUDE_TAGS and CK_EXCLUDE_TAGS work as expected. This is good. I think that there should be something to check that CK_INCLUDE_TAGS and CK_EXCLUDE_TAGS work with CK_RUN_SUITE and CK_RUN_CASE. I do not think there is something for this yet.

cdentyou replied:
Ah yes - true - I will add something.

brarcher commented on src/check.h.in:
```diff
@@ -987,6 +1000,33 @@ CK_DLL_EXP void CK_EXPORT srunner_run(SRunner * sr, const char *sname,
 
 
 /**
+ * Run a specific suite or test case or testcases with specific tags
```
You will need to update the comments on the srunner_run_all and srunner_run functions above to include the following:
```
 * Note that if the CK_RUN_CASE, CK_RUN_SUITE, CK_INCLUDE_TAGS,
  * and/or CK_EXCLUDE_TAGS environment variables are defined, then only
  * the selected suite or test cases will run.
```
namely:
```
 /**
  * Runs a suite runner and all contained suite, printing results to
  * stdout as specified by the print_mode.
  *
  * In addition to running all suites, if the suite runner has been
  * configured to output to a log, that is also performed.
  *
  * Note that if the CK_RUN_CASE, CK_RUN_SUITE, CK_INCLUDE_TAGS,
  * and/or CK_EXCLUDE_TAGS environment variables are defined, then only
  * the selected suite or test cases will run.
  *
  * @param sr suite runner to run all suites from
  * @param print_mode the verbosity in which to report results to stdout
  *
  * @since 0.6.0
  */
 CK_DLL_EXP void CK_EXPORT srunner_run_all(SRunner * sr,
                                           enum print_output print_mode);

 /**
  * Run a specific suite or test case from a suite runner, printing results
  * to stdout as specified by the print_mode.
  *
  * In addition to running any applicable suites or test cases, if the
  * suite runner has been configured to output to a log, that is also
  * performed.
  *
  * Note that if the CK_RUN_CASE, CK_RUN_SUITE, CK_INCLUDE_TAGS,
  * and/or CK_EXCLUDE_TAGS environment variables are defined, then only
  * the selected suite or test cases will run.
  *
  * @param sr suite runner where the given suite or test case must be
  * @param sname suite name to run. A NULL means "any suite".
  * @param tcname test case name to run. A NULL means "any test case"
  * @param print_mode the verbosity in which to report results to stdout
  *
  * @since 0.9.9
  */
 CK_DLL_EXP void CK_EXPORT srunner_run(SRunner * sr, const char *sname,
                                       const char *tcname,
                                       enum print_output print_mode);
```

brarcher replied:
Additionally, the srunner_run_tagged function comment should be updated with this statement.

cdentyou replied:
Good catch.

cdentyou replied:
Slightly reworked your suggested text to make the interaction between the parameters and the env variables explicit.

brarcher commented on doc/check.texi:
```diff
@@ -1314,8 +1316,70 @@ two environment variables available that offers this ability,
+@vindex CK_INCLUDE_TAGS
+@vindex CK_EXCLUDE_TAGS
+
+It can also be useful to be able to dynamically select or exclude
```
It can be useful to dynamically include or exclude groups of tests to be run based on criteria other than the suite or test case name. For example, one or more tags can be assigned to test cases. The tags could indicate if a test runs for a long time, so such tests could be excluded in order to run quicker tests for a sanity check. Alternately, tags may be used to indicate which functional areas test cover. Tests can then be run that include all test cases for a given set of functional areas.

In Check, a tag is a string of characters without white space. One or more tags can be assigned to a test case by using the @code{tcase_set_tags} function. This function accepts a string, and multiple tags can be specified by delimiting them with spaces. For example:
```
Suite *s;
TCase *red, *purple;
s = suite_create("Tag Example");

red = tcase_create("Red");
/* Add the 'Red' tag to the test case */
tcase_set_tags(red, "Red");
suite_add_tcase (s, red);
tcase_add_test(red, red_test1);

purple = tcase_create("Purple");
/* Add both 'Red' and 'Blue' tags to test case */
tcase_set_tags(purple, "Red Blue");
suite_add_tcase (s, purple);
tcase_add_test(purple, purple_test1);
```
Once test cases are tagged they may be selectively run in one of two ways:

a) Using Environment Variables
There are two environment variables available for selecting test cases based on tags: CK_INCLUDE_TAGS and CK_EXCLUDE_TAGS. These can be set to a space separated list of tag names. If CK_INCLUDE_TAGS is set then test cases which include at least one tag in common with CK_INCLUDE_TAGS will be run. If CK_EXCLUDE_TAGS is set then test cases with one tag in common with CK_EXCLUDE_TAGS will not be run. In cases where both CK_INCLUDE_TAGS and CK_EXCLUDE_TAGS match a tag for a test case the test will be excluded.

Both CK_INCLUDE_TAGS and CK_EXCLUDE_TAGS can be specified in conjunction with CK_RUN_SUITE or even CK_RUN_CASE in which case they will have the effect of further narrowing the selection.

b) Programmatically
The @code{srunner_run_tagged} function allows one to specify which tags to run or exclude from a suite runner. This can be used to programmatically control which test cases may run.

cdentyou replied:
Yeah - nicer and more complete - will steal.

brarcher commented on doc/check.texi:
```diff
@@ -974,6 +974,7 @@ easier for the developer to write, run, and analyze tests.
 * Test Fixtures::               
 * Multiple Suites in one SRunner::  
 * Selective Running of Tests::
+* Selecting Tests Based on Arbitrary Tags::
```
You'll need to add:
```
* Selecting Tests by Suite or Test Case
```
here. Otherwise, the index is the following:
```
...
4.5 Multiple Suites in one SRunner	  	
4.6 Selective Running of Tests	  	
4.6.2 Selecting Tests Based on Arbitrary Tags	  	
4.7 Testing Signal Handling and Exit Values	
...  
```

cdentyou replied:
Good catch

brarcher commented:
The merge request is accepted. Thanks for pushing this feature through!

There is a file in Check which lists the Authors of the project: [AUTHORS](https://github.com/libcheck/check/blob/master/AUTHORS). You are welcome to add a merge request for this file, adding yourself under the Contributors list.

cdentyou commented:
Hi Branden

That's great news ! It's been an absolute pleasure working with you and I'm very grateful for all the work you had to put in to get the code into the right shape.  So a huge thank you ! 

Cris

Sent from my iPhone

> On 12 Jul 2016, at 16:47, Branden Archer <notifications@github.com> wrote:
> 
> The merge request is accepted. Thanks for pushing this feature through!
> 
> There is a file in Check which lists the Authors of the project: AUTHORS. You are welcome to add a merge request for this file, adding yourself under the Contributors list.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

brarcher commented:
No problem. Thanks for your contribution. (:
